.github/CONTRIBUTING.md

```diff
@@ -53,68 +57,86 @@ issue. Stale issues will be closed.
-The `go test -short` flag can also be used to skip slower tests.
+### Modifying the Code
 
-Examples (run from the repository root):
-- `go test -v ./connect` will run all tests in the connect package (see `./connect` folder)
-- `go test -v -run TestRetryJoin ./command/agent` will run all tests in the agent package (see `./command/agent` folder) with name substring `TestRetryJoin`
+#### Code Formatting
 
-When a pull request is opened CI will run all tests and lint to verify the change.
+Go provides [tooling to apply consistent code formatting](https://golang.org/doc/effective_go#formatting).
+If you make any changes to the code, run `gofmt -s -w` to automatically format the code according to Go standards.
 
-## Go Module Dependencies
+#### Updating Go Module Dependencies
 
 If a dependency is added or change, run `go mod tidy` to update `go.mod` and `go.sum`.
 
-## Developer Documentation
+#### Developer Documentation
 
-Documentation about the Consul code base is under [./docs],
+Developer-focused documentation about the Consul code base is under [./docs],
 and godoc package document can be read at [pkg.go.dev/github.com/hashicorp/consul].
 
 [./docs]: ../docs/README.md
 [pkg.go.dev/github.com/hashicorp/consul]: https://pkg.go.dev/github.com/hashicorp/consul
 
-### Checklists
+### Testing
+
+During development, it may be more convenient to check your work-in-progress by running only the tests which you expect to be affected by your changes, as the full test suite can take several minutes to execute. [Go's built-in test tool](https://golang.org/pkg/cmd/go/internal/test/) allows specifying a list of packages to test and the `-run` option to only include test names matching a regular expression.
+The `go test -short` flag can also be used to skip slower tests.
+
+Examples (run from the repository root):
+- `go test -v ./connect` will run all tests in the connect package (see `./connect` folder)
+- `go test -v -run TestRetryJoin ./command/agent` will run all tests in the agent package (see `./command/agent` folder) with name substring `TestRetryJoin`
 
-Some common changes that many PRs require such as adding config fields, are
-documented through checklists.
+When a pull request is opened CI will run all tests and lint to verify the change.
 
-Please check in [docs/](../docs/) for any `checklist-*.md` files that might help
-with your change.
+### Submitting a Pull Request
+
+Before writing any code, we recommend:
+- Create a Github issue if none already exists for the code change you'd like to make.
+- Write a comment on the Github issue indicating you're interested in contributing so
+maintainers can provide their perspective if needed.
+
+When you're ready to submit a pull request:
+1. Review the [list of checklists](#checklists) for common changes and follow any
+   that apply to your work.
+2. Include evidence that your changes work as intended (e.g., add/modify unit tests).
```

Comment:
Note a blocker: Another thing that is useful here is to document if they did any manual testing. Often we see people say "We ran this in our dev environment" or something along those lines.

Reply:
Will add that to the `e.g.,` list